juju/juju
# Validate Placement Directives for enable-ha When Spaces Are Configured

## Description of change

Space configuration is automatically supplied as constraints to enable-ha. However, using placement (_juju enable-ha --to x, y_) will happily deploy to machines that cannot access configured spaces, as placement takes precedence over constraints.

This change ensures that when placements are used and there are configured spaces, then the identified machines are checked to ensure they have addresses in the space.

## QA steps

Invalid
1) Prepare 3 nodes on MaaS, with addresses in a specific space.
2) Bootstrap using the space constraint.
3) Add 2 machines ensuring that they *not* in the space.
4) Check that _juju enable-ha --to 1, 2_ returns a validation error.

Valid
1) Repeat steps 1 and 2 above.
2) Add 2 machines using the space constraint.
3) Check that _juju enable-ha --to 1, 2_ works correctly.

With Node Names
1) Repeat steps 1 and 2 above.
3) Check that _juju enable-ha --to {node1}, {node2}_ proceeds without error.

## Documentation changes

Behaviour should be documented against the _enable-ha_ command.

## Bug reference

N/A
